akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala

```diff
@@ -24,6 +26,10 @@ private[akka] trait ReceiveTimeout { this: ActorCell ⇒
 
   final def setReceiveTimeout(timeout: Duration): Unit = receiveTimeoutData = receiveTimeoutData.copy(_1 = timeout)
 
+  protected def checkReceiveTimeoutIfNeeded(message: Any): Unit =
+    if (hasTimeoutData)
+      checkReceiveTimeout(!message.isInstanceOf[NotInfluenceReceiveTimeout])
+
   final def checkReceiveTimeout(reschedule: Boolean = true): Unit = {
     val (recvtimeout, task) = receiveTimeoutData
```

Comment:
I'm surprised that instanceof would have such big impact, could it be and the subsequent checks that is the culprit?

Reply:
It's probably the whole constructing and deconstructing of the tuple. It's a bit hard to say exactly from looking at jmh's `-prof perfasm` output...

The instanceOf calls are definitely not for free, though. The `AutoReceivedMessage` filtering even if never used during the benchmark is quite expensive as well (but I don't see a way around it).

Reply:
I can run as a pattern match 840 times in the time it takes to execute one `isInstanceOf`